src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1120,14 +1120,34 @@ trait Namers extends MethodSynthesis {
 
     /** Computes the type of the body in a ValDef or DefDef, and
      *  assigns the type to the tpt's node.  Returns the type.
+     *
+     *  Under `-Xsource:3`, use `pt`, the type of the overridden member.
+     *  But preserve the precise type of a whitebox macro.
+     *  For `def f = macro g`, here we see `def f = xp(g)` the expansion,
+     *  not the `isMacro` case: `openMacros` will be nonEmpty.
+     *  For `def m = f`, retrieve the typed RHS and check if it is an expansion;
+     *  in that case, check if the expandee `f` is whitebox and preserve
+     *  the precise type if it is. The user must provide an explicit type
+     *  to "opt out" of the inferred narrow type; in Scala 3, they would
+     *  inline the def to "opt in".
      */
     private def assignTypeToTree(tree: ValOrDefDef, defnTyper: Typer, pt: Type): Type = {
       val rhsTpe = tree match {
-        case ddef: DefDef if tree.symbol.isTermMacro => defnTyper.computeMacroDefType(ddef, pt)
+        case ddef: DefDef if tree.symbol.isTermMacro => defnTyper.computeMacroDefType(ddef, pt) // unreached, see methodSig
```

Comment:
So this can be removed?

There's also a comment in `computeMacroDefType` saying

> if a macro def doesn't have explicitly specified return type, this method will be called again by `assignTypeToTree`

is that stale?

Reply:
I was reluctant to expand the scope, as macros remain experimental.